telemetry/server.go

```diff
@@ -57,109 +56,72 @@ func handleHealthCheck(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprintf(w, "OK")
 }
 
-type TelemetryType string
-
-const (
-	ProtocolStatsMetric        TelemetryType = "ProtocolStats"
-	ReceivedEnvelopeMetric     TelemetryType = "ReceivedEnvelope"
-	SentEnvelopeMetric         TelemetryType = "SentEnvelope"
-	UpdateEnvelopeMetric       TelemetryType = "UpdateEnvelope"
-	ReceivedMessagesMetric     TelemetryType = "ReceivedMessages"
-	ErrorSendingEnvelopeMetric TelemetryType = "ErrorSendingEnvelope"
-	PeerCountMetric            TelemetryType = "PeerCount"
-	PeerConnFailureMetric      TelemetryType = "PeerConnFailure"
-)
-
-type TelemetryRequest struct {
-	Id            int              `json:"id"`
-	TelemetryType TelemetryType    `json:"telemetry_type"`
-	TelemetryData *json.RawMessage `json:"telemetry_data"`
-}
-
 func (s *Server) createTelemetryData(w http.ResponseWriter, r *http.Request) {
 	start := time.Now()
-	var telemetryData []TelemetryRequest
+	var telemetryData []types.TelemetryRequest
 	decoder := json.NewDecoder(r.Body)
 	if err := decoder.Decode(&telemetryData); err != nil {
 		log.Println(err)
 		http.Error(w, "Failed to decode telemetry data", http.StatusBadRequest)
 		return
 	}
 
-	var errorDetails []map[string]interface{}
+	var errorDetails MetricErrors
 
 	for _, data := range telemetryData {
 		switch data.TelemetryType {
-		case ProtocolStatsMetric:
+		case types.ProtocolStatsMetric:
 			var stats ProtocolStats
-			if err := json.Unmarshal(*data.TelemetryData, &stats); err != nil {
-				errorDetails = append(errorDetails, map[string]interface{}{"Id": data.Id, "Error": fmt.Sprintf("Error decoding protocol stats: %v", err)})
+			err := stats.process(s.DB, &errorDetails, &data)
+			if err != nil {
```

Comment:
Should we log the err? Same question for lines 81 87 93 99 105 and 111, or is the log in line 120 enough?

Reply:
I could also add logging the error to the `errors.append` method to make sure every new error is logged?

Reply:
Tried, this would result in telemetry logs like the following, and we could remove the log on line 120 to not duplicate

```
2024-08-08T10:28:43.752+0200	ERROR	telemetry/errors.go:28	Error saving peer count: pq: duplicate key value violates unique constraint "peercount_unique"
github.com/status-im/dev-telemetry/telemetry.(*MetricErrors).Append
	/home/vpavlin/devel/github.com/vpavlin/telemetry/telemetry/errors.go:28
github.com/status-im/dev-telemetry/telemetry.(*PeerCount).process
	/home/vpavlin/devel/github.com/vpavlin/telemetry/telemetry/peer_count.go:41
github.com/status-im/dev-telemetry/telemetry.(*Server).createTelemetryData
	/home/vpavlin/devel/github.com/vpavlin/telemetry/telemetry/server.go:99
net/http.HandlerFunc.ServeHTTP
	/usr/lib/go/src/net/http/server.go:2136
github.com/rs/cors.(*Cors).Handler-fm.(*Cors).Handler.func1
	/home/vpavlin/devel/pkg/mod/github.com/rs/cors@v1.11.0/cors.go:289
net/http.HandlerFunc.ServeHTTP
	/usr/lib/go/src/net/http/server.go:2136
github.com/status-im/dev-telemetry/telemetry.(*Server).rateLimit-fm.(*Server).rateLimit.func1
	/home/vpavlin/devel/github.com/vpavlin/telemetry/telemetry/server.go:222
net/http.HandlerFunc.ServeHTTP
	/usr/lib/go/src/net/http/server.go:2136
github.com/gorilla/mux.(*Router).ServeHTTP
	/home/vpavlin/devel/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210
net/http.serverHandler.ServeHTTP
	/usr/lib/go/src/net/http/server.go:2938
net/http.(*conn).serve
	/usr/lib/go/src/net/http/server.go:2009
```